Let's use some Twilio Studio Flows!

katrinamariehh opened the pull request:
For [this card](https://www.pivotaltracker.com/story/show/157493533) I wrote a function to trigger a studio flow for a single phone number + state and another function that can take in a list of phone numbers + states and trigger the studio flow for all of them. I did the beginning process of configuring the ideas42 Twilio sid and token to get used here; to keep it basic, triggering a studio flow will require the SID for the flow and the associated `from-number` to use (for this to work properly a number needs to be configured with a specific flow so each flow will have a unique `from-number`). For more long-term use we would also want to abstract out the extra parameter (in this case `state_abbrev`) but it's used in both of the flows we want to test so I'd rather have it in there now for potential testing later this week and figure out something more generic later.

theJohnnyBrown commented on src/sms_works/handlers.clj:
```diff
@@ -28,3 +30,23 @@
+        _ (log/debug "Starting studio flow with:" (str phone-number))
+        result (client/post (str "https://studio.twilio.com/v1/Flows/" flow-sid "/Engagements")
+                            {:basic-auth [(config [:twilio :ideas42-sid]) (config [:twilio :ideas42-token])]
+                             :form-params data})]
```
It feels weird to me having form params with one of the values being a json string. It would be better with `:content-type :json` and letting `client` serialize the body.

gws replied:
This is, for better or worse, how the Twilio Studio API works.

mrichards42 replied:
It's looking for a json string here, which clj-http won't handle:

From <https://www.twilio.com/docs/studio/user-guide#rest-api>
> Note: the JSON value must explicitly be passed as a string, not as a hash object

mrichards42 replied:
beat me to it 😃

theJohnnyBrown commented on src/sms_works/handlers.clj:
```diff
@@ -28,3 +30,23 @@
+  "Starts the specific Studio Flow based on flow-sid
+   and from-number for a given phone number via Twilio"
+  [phone-number state flow-sid from-number]
+  (let [data {:From from-number
```
Would be nice to see some error logging around the twilio request

katrinamariehh replied:
What kind of logging were you thinking? Just like checking the status on the response and logging that it wasn't successful? Something like that?

katrinamariehh replied:
Are you suggesting a more general change to `sms-works` or just sticking some error-checking in with what I have for now? I just want to make sure I understand what is going to work with what we're trying to accomplish here vs. long-term changes we should make in this service.

theJohnnyBrown replied:
For repl use, I actually found it nice that errors bubbled up to where cider could display them with highlighting and everything, so for that kind of thing it's good how it is.

For use as a handler in production, I would want to see logging of non-200 status codes (which you've added) and something to catch any exceptions and log/error them.

Most of the services wrap their handler functions in something like [user-works's error handler](https://github.com/democracyworks/user-works/blob/master/src/user_works/handlers.clj#L24) (which should probably be factored out into our shared turbovote library).

gws commented on src/sms_works/handlers.clj:
```diff
@@ -28,3 +30,23 @@
                  (twilio/send-sms message))]
     (log/debug "Twilio result:" (pr-str @result))))
+
+(defn trigger-studio
```
Typically the things in `handlers` namespaces are actually handler functions for requests/messages coming in from outside the system (messages from a RabbitMQ queue delivered via Kehaar, for instance). Where will these requests originate from? From what I can tell, a developer will trigger this flow via the REPL or something, is that right?

katrinamariehh replied:
Yeah, for the initial effort we're just going to trigger this via a REPL -- long-term we may want formalize it more and I'd be happy to re-name it then to make more sense with the other handler functions.

gws replied:
OK, I understand the balance you're striking here since this is kind of an MVP. Were you to have started this from scratch, I might suggest putting all of this in `dev-src` (where we put REPL/CLI developer-driven code) until it was ready to be promoted to an actual handler, but since there isn't much here and that's asking a lot, I'll just leave it at that. If you'd like to add a comment documenting the intent you had for these two functions, that might be nice. :+1:

katrinamariehh replied:
👍🏻

katrinamariehh commented on src/sms_works/handlers.clj:
```diff
@@ -2,7 +2,9 @@
-            [twilio.core :as twilio]))
+            [twilio.core :as twilio]
+            [clj-http.client :as client]
+            [cheshire.core :refer :all]))
```
Cool, I can definitely change this.

katrinamariehh commented on dev-src/sms_works/studio_flow.clj:
```diff
@@ -0,0 +1,32 @@
+                             :form-params data})]
+    (if (= (:status result) 200)
+      (do
+        (log/debug "Engagement sid: " (:sid (json/parse-string (:body result) true)))
```
ah yeah, this would definitely be a better way to do it than just dumping stuff out --  I'll for sure keep this in mind as we revise it!

katrinamariehh commented on dev-src/sms_works/studio_flow.clj:
```diff
@@ -0,0 +1,32 @@
+  (let [data {:From from-number
+              :To phone-number
+              :Parameters (json/generate-string {:state_abbrev state})}
+        _ (log/debug "Executing studio flow:" (str flow-sid))
```
honestly I did it because I think I saw it somewhere else? I'm happy to move it, though! And thanks for giving me an example of how to fix it!

katrinamariehh commented on dev-src/sms_works/studio_flow.clj:
```diff
@@ -0,0 +1,32 @@
+        (log/debug "Engagement sid: " (:sid (json/parse-string (:body result) true)))
+        (log/debug "Response body:" (json/parse-string (:body result) true)))
+      (do
+        (log/debug "Did not get a 200")
```
That makes sense -- we want to send about 1000 of each message out, let's touch base in the morning if this is maybe a prohibitive concern (i.e. if 1000 messages is definitely going to fill the queue and cause problems that we don't have a way to handle right now).

katrinamariehh commented:
Ok! So anyone who's still interested -- this was used last week successfully! I refactored it so that when you trigger studio flows for a set of users it outputs a file with information about whether or not it succeeded and added a piece that would take that output file and generate another file with user responses. 

Things that I'm not fixing yet:
- still specifically expects "registeredaddressstate" which is needed for the two existing flows that we have, could be generalized to take in any number of extra columns and turn them into parameters to pass on
- instead of putting all output into a single file, it would make sense to generate one file with successful calls and one with the failed calls to make it easier to re-try the ones that failed

I also tried to clean it up style-wise, I'd be happy for any feedback about how I attempted to break up my long lines.

mrichards42 commented on dev-src/sms_works/studio_flow.clj:
```diff
@@ -0,0 +1,76 @@
+        headers [["lastname" "firstname" "phone" "id" "registeredaddressstate" "flowsid" "body" "date"]]]
+    (with-open [writer (io/writer output-filename)]
+      (csv/write-csv writer headers)
+      (csv/write-csv writer (apply concat output-data)))))
```
A couple of notes:

You should be able to use [`seq`](https://clojuredocs.org/clojure.core/seq) in place of `#(not= (count %) 0)`. Seq returns `nil` for empty sequences.


From reading through this code, it looks like the shape of the `output` data is a sequence of vectors. I don't think you need to filter it at all, since as a result of using `concat`, the empty vectors are removed:

```clojure
=> (concat [[1 2 3] [4 5 6]] [] [[7 8 9]])
([1 2 3] [4 5 6] [7 8 9])
```
You could also use [`mapcat`](https://clojuredocs.org/clojure.core/mapcat), which is essentially `(apply concat (map ...))`. If you don't need the filter, the map/filter/concat code could be replaced with:
```clojure
(let [output (mapcat (partial ...)
                     (rest rows))]
  ...
  (csv/write-csv writer output))
```
